kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,49 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks.
+    /// This callback is called when a digest is computed.
+    /// The callback should follow the `Client` `hash_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes written on success
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8], length: usize) -> Result<usize, ReturnCode>;
+
+    /// Request the hardware block to generate a Digest
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self);
+
+    /// Stores the returned digest in the memory location specified. This
+    /// is the only way to get the output of the digest.
+    fn get_digest(&self, digest: &mut [u8], len: usize) -> ReturnCode;
+
+    /// Clear the keys and any other sensitive data.
+    /// This won't clear the buffers provided to this API, that is up to the
+    /// user to clear.
+    fn clear_data(&self);
+}
+
+pub trait HMACSha256 {
+    /// Call before `Digest::run()` to perform HMACSha256
+    ///
+    /// The key used for the HMAC is passed to this function as well
+    /// as the length of the key data.
+    fn set_mode_hmacsha256(&self, key: &[u8], len: usize);
```

Comment:
Should this be a fixed, known keysize (i.e., &[u8; 16])?

Reply:
HMAC doesn't limit the size of the key. If it's bigger than a block for the selected underlying digest, the provided key is hashed before being used

Reply:
But providing the length is redundant. The slice already embeds it.

Reply:
I have left this as not a known size and removed the length argument.